[FLINK-12989][hive] Generate HiveTableSink from from a Hive table

xuefuz opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](http://flink.apache.org/contribute-code.html#best-practices).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Add the generation of HiveTableSink from from a Hive table

## Brief change log

*(for example:)*
  - Added the logic of generation of HiveTableSink from from a Hive table
  - Refactored and simplified HiveTableSink related code
  - Added test for the logic

## Verifying this change

This change is verified with existing and new test cases.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yno)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (yes )
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

xuefuz commented:
Please note that this PR included the changes in PR #8883. cc: @bowenli86

zjuwangg commented:
better to rename the pr title

lirui-apache commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ObjectPath.java:
```diff
@@ -28,8 +31,19 @@
 public class ObjectPath {
-	private final String databaseName;
-	private final String objectName;
+	// Default database name
```
How about just make ObjectPath a Serializable?

lirui-apache commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogTableImpl.java:
```diff
@@ -32,65 +32,39 @@
 	}
 
 	public CatalogTableImpl(
+			ObjectPath tablePath,
```
If a CatalogTable can have an ObjectPath field, do we need some validation in APIs like `HiveCatalog::createTable`?
And btw, why did we decide to separate CatalogTable and ObjectPath in the first place?

xuefuz replied:
I added some validation.

As to separation, CatalogTable as API doesn't require ObjectPath, as it's consistent how objects are organized in databases. However, as an impl class such as CatalogTableImpl, it can has anything it needs without violating the API. One implementation might not need this, and we don't want to enforce this at the API level.

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableFactory.java:
```diff
@@ -31,13 +32,22 @@
  * A table factory implementation for tables stored in Hive catalog.
  */
 public class HiveTableFactory implements TableSourceFactory<Row>, TableSinkFactory<Row> {
+	private HiveCatalog hiveCatalog;
```
True. In that case, I think we can just pass on HiveConf instead.

xuefuz commented:
Based on feedback, I have changed the design a bit so we don't store table path in catalog table any more. Instead, we added ObjectPath in TableSource/SinkFactory APIs. This simplified and reduced the changes.

@lirui-apache @bowenli86 pls take another look. thanks.

lirui-apache commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableSink.java:
```diff
@@ -55,30 +58,31 @@
 
-	public HiveTableSink(JobConf jobConf, RowTypeInfo rowTypeInfo, String dbName, String tableName,
-			List<String> partitionColumns) {
+	public HiveTableSink(JobConf jobConf, ObjectPath tablePath, CatalogTableImpl table) {
```
Seems we just need a CatalogTable here too

xuefuz commented:
> @xuefuz thanks for the update!
> 
> Just want to confirm if the approach also works for in-memory catalog, because currently users can create Hive tables in in-memory catalog. Would be better to work for both HiveCatalog (HiveTableFactory) and in-memory catalog. If we need extra work for it, it can be in the different PR.
> 
> Otherwise LGTM with a few minor comments.

Good point. I expect that GenericInMemoryCatalog will have its own table factory, which will need to process the request according to the actual table type in a similar fashion as in HiveTableFactory. However, that's a different thing with lower priority. We can do that as a followup.

xuefuz commented on flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/batch/connectors/hive/HiveTableSinkTest.java:
```diff
@@ -113,8 +114,8 @@ public void testInsertIntoDynamicPartition() throws Exception {
-		tableEnv.registerTableSink("destSink", new HiveTableSink(new JobConf(hiveConf), rowTypeInfo,
-				"default", "dest", HiveCatalog.getFieldNames(hiveTable.getPartitionKeys())));
+		CatalogBaseTable table = hiveCatalog.getTable(tablePath);
+		tableEnv.registerTableSink("destSink", new HiveTableSink(new JobConf(hiveConf), tablePath, (CatalogTableImpl) table));
```
HiveTableSink constructor requires a CatalogTableImpl instance.

xuefuz commented on flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/batch/connectors/hive/HiveTableSinkTest.java:
```diff
@@ -190,8 +191,8 @@ public void testWriteComplexType() throws Exception {
 		hiveTable = hiveCatalog.getHiveTable(tablePath);
-		tableEnv.registerTableSink("nestedSink", new HiveTableSink(new JobConf(hiveConf), rowTypeInfo,
-				"default", "dest", HiveCatalog.getFieldNames(hiveTable.getPartitionKeys())));
+		catalogTable = (CatalogTableImpl) hiveCatalog.getTable(tablePath);
```
Same as above.

bowenli86 commented:
Merging
